### ISSUE TITLE
Instantiating a mapped type does not update the constraint on the type parameter

### ISSUE THREAD

**Issue by mattmccutchen** [Microsoft/TypeScript#27596]
I don't expect real users to write code like this example.  I just wanted to get this bug out of the way because I think it will make the solution to #27484 easier to understand.

**TypeScript Version:**  master (85a3475df87b9c10240570f32090970262c99b54)

**Search Terms:** mapped type parameter constraint instantiate

**Code**
```ts
type MyMap<T> = {[P in keyof T]: T[keyof T]};
function foo<U>(arg: U): MyMap<U> {
    return arg;
}
```

**Expected behavior:** No error.

**Actual behavior:** Error:
```
Type 'U' is not assignable to type 'MyMap<U>'.
  Type 'U[P]' is not assignable to type 'U[keyof U]'.
    Type 'P' is not assignable to type 'keyof U'.
      ...
```
This is nonsense because `P in keyof T` and we set `T = U`.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=type%20MyMap%3CT%3E%20%3D%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20T%5Bkeyof%20T%5D%20%7D%0D%0Afunction%20oops%3CU%3E(arg%3A%20U)%3A%20MyMap%3CU%3E%20%7B%20%0D%0A%20%20%20%20return%20arg%3B%0D%0A%7D%0D%0A)

**Related Issues:** none found
